### PR TITLE
Zephyr module variables in Kconfig

### PR DESCRIPTION
**Pull request by tejlmand**
With the introduction of https://github.com/zephyrproject-rtos/zephyr/pull/26715 it became apparent that downstream users of Zephyr were dependent on `SOC_DIR` variable in Kconfig.

Because of this, and to align Kconfig and CMake Zephyr module variables between Kconfig and CMake, then this PR proposes to export `ZEPHYR_<module>_MODULE_DIR` to Kconfig.

In CMake today, if module `FOO` is part of Zephyr, then the following is possible in CMake:
```
include(${ZEPHYR_FOO_MODULE_DIR}/myfile.cmake)
```

with this PR, the similar is now possible in Kconfig:
```
source "$(ZEPHYR_FOO_MODULE_DIR)/Kconfig.myfile"
```

This means that users who previously did:
```
source "$(SOC_DIR)/$(ARCH)/Kconfig.mysoc"
```

can now do:
```
source "$(ZEPHYR_MYSOC_MODULE_DIR)/soc/$(ARCH)/Kconfig.mysoc"
```
if the repo `MYSOC` is also a Zephyr module.

--------

Edit: Based on comments, then additional commits has been added that does general clean-up of CMake/python code related to Zephyr modules.

### CODE REVIEW

**Comment by tejlmand** [comment]
@dkalowsk FYI

File: doc/guides/modules.rst [hunk 1]
```diff
@@ -75,6 +75,25 @@ module:
...
      kconfig: Kconfig
 
+
+Build system integration
+========================
+
+When a module has been described using the :file:`module.yml` file then it will
+automatically be included into the Zephyr Kconfig and CMake build system.
+
+To facilitate referencing between Zephyr and Zephyr modules, then all modules
```
Comment (dkalowsk) [inline doc/guides/modules.rst:85]: minor edit suggestion, drop the `then`.

File: cmake/kconfig.cmake [hunk 2]
```diff
@@ -63,6 +63,20 @@ set(ENV{ARCH_DIR}   ${ARCH_DIR})
...
+# source "$(ZEPHYR_FOO_MODULE_DIR)/Kconfig"
+foreach(module_name ${ZEPHYR_MODULE_NAMES})
+  string(TOUPPER ${module_name} MODULE_NAME_UPPER)
+  list(APPEND
+       ZEPHYR_KCONFIG_MODULES_DIR
+       "ZEPHYR_${MODULE_NAME_UPPER}_MODULE_DIR=${ZEPHYR_${MODULE_NAME_UPPER}_MODULE_DIR}"
+  )
+  set(ENV{ZEPHYR_${MODULE_NAME_UPPER}_MODULE_DIR}
+      ${ZEPHYR_${MODULE_NAME_UPPER}_MODULE_DIR}
+  )
```
Comment (mbolivar-nordic) [inline cmake/kconfig.cmake:77]: Why is this necessary to set in the main build environment? I would guess adding it to the environments of each kconfig_target would be enough. We have too many environment variables already IMO.

File: doc/guides/modules.rst [hunk 3]
```diff
@@ -75,6 +75,25 @@ module:
...
+will be accessible through Kconfig and CMake variables.
+
+In Kconfig and CMake, the ``ZEPHYR_<module-name>_MODULE_DIR`` can be used to
```
Comment (mbolivar-nordic) [inline doc/guides/modules.rst:88]: I didn't raise a bug about this because it wasn't officially documented before, but this is not really correct.

The existence of a module.yml file does not guarantee this variable is defined, even in CMake.

See this source code comment in one of my PRs for details: https://github.com/zephyrproject-rtos/zephyr/pull/27688/files#diff-a1f9e35746e0274b9ad8cfb12162414dR38

I think this situation should be fixed, or at least documented more clearly.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]

File: cmake/kconfig.cmake (hunk 2 above)
Comment (mbolivar-nordic) [reply cmake/kconfig.cmake:77]: (This one is not a blocker, just asking if we can avoid adding extra environment variables to the cmake process.)

**Comment by dkalowsk** [comment]
I don't have the ability to fully test out the PR within my current setup as my SOC_DIR is not a module, but making some quick changes the proposed work flow does appear to work.  AKA I can boot on a UUT with the changes.  

My only real comment on this change is that the PR cover documentation.  I don't know what the documentation process is like for Zephyr so feel free to ignore this.  The PR cover documentation gives a little more detail than what the proposed included documentation changes have.  As someone new coming to setup, it would be really useful to have those details included in the documentation, and the reason for why this method is better than another method.

File: doc/guides/modules.rst (hunk 3 above)
Comment (tejlmand) [reply doc/guides/modules.rst:88]: Agreed, as minimum this should be rephrased to be more precise that only if build files are present in the module.

Guess the introduction of sanity in `module.yml` made a small hole here.

File: cmake/kconfig.cmake (hunk 2 above)
Comment (tejlmand) [reply cmake/kconfig.cmake:77]: I completely agree, however this commit follows the existing lines here:
https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/kconfig.cmake#L51-L64
and is currently needed because an `execute_process()` here:
https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/kconfig.cmake#L213-L216

Cleaning up this is outside scope of this PR, but con follow in a dedicated PR covering all above mentioned lines.

**Comment by tejlmand** [comment]
> As someone new coming to setup, it would be really useful to have those details included in the documentation, and the reason for why this method is better than another method.

That's a good point. I will try to add more detailed example, but I will try to avoid referring `SOC_DIR` in the docs.
Reason for that, is that new comers to Zephyr should not be confused about `SOC_DIR` as it is no longer valid.
(Such info belongs to the release notes.)

But for reviewers of this PR, I believe it is of importance to know the motivation for this enhancement, and there the `SOC_DIR` example is of importance.

File: doc/guides/modules.rst (hunk 3 above)
Comment (mbolivar-nordic) [reply doc/guides/modules.rst:88]: > as minimum this should be rephrased to be more precise that only if build files are present in the module.

I agree that would be the minimum acceptable change, since it makes the documentation correct.

Comment (mbolivar-nordic) [reply doc/guides/modules.rst:88]: (Although FTR I would appreciate it if the existence of a module.yml resulted in the definition of ZEPHYR_FOO_MODULE_DIR. That is the best possible world but I agree it's outside the scope of this PR)

File: cmake/kconfig.cmake (hunk 2 above)
Comment (mbolivar-nordic) [reply cmake/kconfig.cmake:77]: > and is currently needed because an `execute_process()` here:
> https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/kconfig.cmake#L213-L216

Got it. Thanks for the reference.

**Comment by tejlmand** [comment]
> As someone new coming to setup, it would be really useful to have those details included in the documentation, and the reason for why this method is better than another method.

@ dkalowsk Done, example updated.
Feedback is welcome.

File: doc/guides/modules.rst (hunk 3 above)
Comment (tejlmand) [reply doc/guides/modules.rst:88]: couldn't leave code like that.
Fixed. Now docs and code is aligned.

File: cmake/kconfig.cmake (hunk 2 above)
Comment (tejlmand) [reply cmake/kconfig.cmake]: included a cleanup commit.

File: doc/guides/modules.rst (hunk 1 above)
Comment (tejlmand) [reply doc/guides/modules.rst]: fixed

File: doc/guides/modules.rst (hunk 3 above)
Comment (mbolivar-nordic) [reply doc/guides/modules.rst:88]: OK, rebased and tested on top of the updated version. I confirm the fix. Thanks for fixing :+1:

File: doc/guides/modules.rst [hunk 4]
```diff
@@ -75,6 +75,65 @@ module:
...
+
+- Module directory in CMake: ``${ZEPHYR_FOO_MODULE_DIR}``
+- Module directory in Kconfig: ``$(ZEPHYR_FOO_MODULE_DIR)``
+- :file:`CMakeLists.txt` directory include in build system: ``${ZEPHYR_FOO_CMAKE_DIR}``
+
+Those variables may be used to obtain a full path to Zephyr module or to test
+whether a given module is present in the system.
+
+For example, to verify that ``FOO`` exists in the system and is a Zephyr module
+then one may do:
```
Comment (mbolivar-nordic) [inline doc/guides/modules.rst:105]:
```suggestion
Notice how a lowercase module name ``foo`` is capitalized to ``FOO``
in both CMake and Kconfig.

These variables can also be used to test whether a given module exists.
For example, to verify that ``foo`` is the name of a Zephyr module:
```

File: doc/guides/modules.rst [hunk 5]
```diff
@@ -75,6 +75,65 @@ module:
...
+
+  if(ZEPHYR_FOO_MODULE_DIR)
+    # Do something if FOO exists.
+  endif()
+
+In Kconfig, the variable may be used to specify a path to include.
+An example can be inclusion of additional Kconfig SoC files.
+
+In this example ``FOO`` specifies an additional ``SOC_ROOT`` and in
+:file:`<FOO>/soc/arm/foo_soc/Kconfig.soc` one may do:
```
Comment (mbolivar-nordic) [inline doc/guides/modules.rst:117]:
```suggestion
In Kconfig, the variable may be used to find additional files to include.
For example, to include the file :file:`some/Kconfig` in module ``foo``:
```

File: doc/guides/modules.rst [hunk 6]
```diff
@@ -75,6 +75,65 @@ module:
...
+
+.. code-block:: kconfig
+
+  source "$(ZEPHYR_FOO_MODULE_DIR)/soc/arm/foo_soc/Kconfig.soc.common"
```
Comment (mbolivar-nordic) [inline doc/guides/modules.rst:121]:
```suggestion
  source "$(ZEPHYR_FOO_MODULE_DIR)/some/Kconfig"
```

File: doc/guides/modules.rst [hunk 7]
```diff
@@ -75,6 +75,65 @@ module:
...
+
+During CMake processing of each Zephyr module, the following two variables are
+also available:
+
+- Module directory in CMake of current module being processed: ``${ZEPHYR_CURRENT_MODULE_DIR}``
+- :file:`CMakeLists.txt` directory of current module directory: ``${ZEPHYR_CURRENT_CMAKE_DIR}``
```
Comment (mbolivar-nordic) [inline doc/guides/modules.rst:127]:
```suggestion
- the current module's top level directory: ``${ZEPHYR_CURRENT_MODULE_DIR}``
- the current module's :file:`CMakeLists.txt` directory: ``${ZEPHYR_CURRENT_CMAKE_DIR}``
```

File: doc/guides/modules.rst [hunk 8]
```diff
@@ -75,6 +75,65 @@ module:
...
+
+This remove the need for a Zephyr module to know its own name during CMake
+processing, and the module can source additional CMake files using the current
+variable, for example:
```
Comment (mbolivar-nordic) [inline doc/guides/modules.rst:131]:
```suggestion
This removes the need for a Zephyr module to know its own name during CMake
processing. The module can source additional CMake files using these ``CURRENT``
variables. For example:
```

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
I have a few cleanups I think would make the docs a bit clearer and then I am +1. Thanks again for taking the time to put together a proper fix.

File: doc/guides/modules.rst (hunk 8 above)
Comment (mbolivar-nordic) [reply doc/guides/modules.rst:131]: @tejlmand it occurs to me that this means no module can be named "current".

Worth enforcing in the build system?

**Comment by dkalowsk** [comment]
It looks like @mbolivar-nordic has already suggested the same changes to the documentation I was in the process of creating, so I won't bother doubling up on those.  Appreciate the effort in cleaning and describing everything for the updated process and details.

File: doc/guides/modules.rst (hunk 8 above)
Comment (tejlmand) [reply doc/guides/modules.rst:131]: Good catch. Let's hope someone doesn't create a current measurement sw tool, and call it `current`.

File: doc/guides/modules.rst [hunk 9]
```diff
@@ -75,6 +75,65 @@ module:
...
+
+To facilitate referencing between Zephyr and Zephyr modules all modules
+will be accessible through Kconfig and CMake variables.
+
+In Kconfig and CMake, the ``ZEPHYR_<module-name>_MODULE_DIR`` can be used to
+obtain the absolute path to the module directory.
+
+In CMake, the ``ZEPHYR_<module-name>_CMAKE_DIR`` can be used to obtain the
+absolute path to the :file:`CMakeLists.txt` that is included into CMake build
+system.
```
[the remark this answers is not in the document]
Comment (tejlmand) [reply doc/guides/modules.rst:91]: Not completely accurate. The  `ZEPHYR_<module-name>_CMAKE_DIR` is only dependent on CMakeLists.txt file.

File: doc/guides/modules.rst (hunk 8 above)
Comment (tejlmand) [reply doc/guides/modules.rst]: added extra commit to check for `current`: https://github.com/zephyrproject-rtos/zephyr/pull/27740/commits/8511f9fd9a854c1963a415a6a5e93681eaf0d79b

**Review by mbolivar-nordic — APPROVED** [review]
Looks great! Thanks again.

File: doc/guides/modules.rst (hunk 8 above)
Comment (mbolivar-nordic) [reply doc/guides/modules.rst]: > added extra commit to check for `current`: [8511f9f](https://github.com/zephyrproject-rtos/zephyr/commit/8511f9fd9a854c1963a415a6a5e93681eaf0d79b)

nice :+1:

File: cmake/zephyr_module.cmake [hunk 10]
```diff
@@ -59,6 +59,32 @@ if(WEST OR ZEPHYR_MODULES)
...
+      # lazy regexes (it supports greedy only).
+      string(REGEX REPLACE "\"(.*)\":\".*\":\".*\"" "\\1" module_name ${module})
+      string(REGEX REPLACE "\".*\":\"(.*)\":\".*\"" "\\1" module_path ${module})
+      string(REGEX REPLACE "\".*\":\".*\":\"(.*)\"" "\\1" cmake_path ${module})
+
+      list(APPEND ZEPHYR_MODULE_NAMES ${module_name})
+
+      string(TOUPPER ${module_name} MODULE_NAME_UPPER)
+      if(NOT ${MODULE_NAME_UPPER} STREQUAL CURRENT)
+        set(ZEPHYR_${MODULE_NAME_UPPER}_MODULE_DIR ${cmake_path})
```
Comment (mbolivar-nordic) [inline cmake/zephyr_module.cmake:79]: I rebased and retested the new version and it looks like a typo snuck in here. This should be:

```suggestion
        set(ZEPHYR_${MODULE_NAME_UPPER}_MODULE_DIR ${module_path})
```

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]

File: cmake/zephyr_module.cmake (hunk 10 above)
Comment (tejlmand) [reply cmake/zephyr_module.cmake]: Good catch, thanks.

Wonder how that sneaked in, cause it is correct here: https://github.com/zephyrproject-rtos/zephyr/pull/27740/commits/67c09760e740f9872275592c99d23ffec7d52036
Guess i'll never know.

**Review by mbolivar-nordic — APPROVED** [review]
Working for me again, thanks for the fast update.

**Comment by carlescufi** [comment]
Does this affect CI? https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/ci/check_compliance.py#L240

**Comment by tejlmand** [comment]
> 
> 
> Does this affect CI? https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/ci/check_compliance.py#L240

No, this PR has no behavioral change on the build system, it only offers additional informational variables that can be used elsewhere in the build system. (and does some cleanup of existing build code)

**Review by carlescufi — APPROVED** [review]
